File: 0316-remove-duplicate-letters/0316-remove-duplicate-letters.java
class Solution {
    public String removeDuplicateLetters(String s) {
        int[] count = new int[26];
        boolean[] visited = new boolean[26];
        Stack<Character> stack = new Stack<>();
        int n = s.length();

        for (char ch : s.toCharArray()) {
            count[ch - 'a']++;
        }

        for (char ch : s.toCharArray()) {
            count[ch - 'a']--;

            if (visited[ch - 'a']) {
                continue;
            }

            while (!stack.isEmpty() && stack.peek() > ch && count[stack.peek() - 'a'] > 0) {
                visited[stack.peek() - 'a'] = false;
                stack.pop();
            }

            stack.push(ch);
            visited[ch - 'a'] = true;
        }

        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.insert(0, stack.peek());
            stack.pop();
        }

        return sb.toString();
    }
}

File: 0316-remove-duplicate-letters/0316-remove-duplicate-letters.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        int n = s.size();
        vector<int> count(26, 0);
        vector<bool> visited(26, false);
        stack<char> bucket;
        for(int i=0;i<n;i++){
            count[s[i]-'a']++;
        }
        for(int i=0;i<n;i++){
            count[s[i]-'a']--;
            if(visited[s[i]-'a']){
                continue;
            }
            while(!bucket.empty() && bucket.top()>s[i] && count[bucket.top()-'a']>0){
                visited[bucket.top()-'a'] = false;
                bucket.pop();
            }
            bucket.push(s[i]);
            visited[s[i]-'a']=true;
        }
        string result = "";
        while(!bucket.empty()){
            result.insert(result.begin(), bucket.top());
            bucket.pop();
        }
        return result;
    }
};